### ISSUE TITLE
~40% performance decrease since Tensorflow 1.9 when training large models

### ISSUE THREAD

**Issue by DavidWiesner** [tensorflow/tensorflow#23847]
**System information**
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 16.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): v1.9.0-0-g25c197e 1.9.0
- Python version: 3
- Bazel version (if compiling from source): -
- GCC/Compiler version (if compiling from source): -
- CUDA/cuDNN version: 9.0.176/7.1.4.18
- GPU model and memory: 8 x Tesla V100-SXM2-16GB major: 7 minor: 0 memoryClockRate(GHz): 1.53

**Describe the current behavior**
I train a model using slim in tensorflow models with the flowers dataset and the large nasnet. This were run on the official tensorflow docker images.
On my Tesla V100 for tensorflow 1.8.0-devel-gpu-py3 I got around 
`INFO:tensorflow:global_step/sec: 1.33334`
With tensorflow 1.9.0-devel-gpu-py3 a maximum at
`INFO:tensorflow:global_step/sec: 0.89946`
With the current tensorflow 1.12.0-devel-gpu-py3 a maximum at
`INFO:tensorflow:global_step/sec: 0.900003`

**Describe the expected behavior**
There should not be an performance decrease in this drastic 

**Code to reproduce the issue**
```shell
#!/usr/bin/env bash
[ ! -d "models" ] && git clone https://github.com/tensorflow/models.git
cd models/research/slim/
MODEL_NAME="nasnet_large"
TRAIN_DIR=/tmp/flowers-models/${MODEL_NAME}
DATASET_DIR=/tmp/flowers
rm -rf "$TRAIN_DIR"
export CUDA_DEVICE_ORDER="PCI_BUS_ID"
export CUDA_VISIBLE_DEVICES="0,1"

python download_and_convert_data.py \
    --dataset_name=flowers \
    --dataset_dir=${DATASET_DIR}

python train_image_classifier.py \
    --train_dir=${TRAIN_DIR}/all \
    --dataset_name=flowers \
    --dataset_split_name=train \
    --dataset_dir=${DATASET_DIR} \
    --model_name=${MODEL_NAME} \
    --max_number_of_steps=500 \
    --batch_size=16 \
    --learning_rate=0.015 \
    --save_interval_secs=60000 \
    --save_summaries_secs=60 \
    --log_every_n_steps=10 \
    --optimizer=rmsprop \
    --num_preprocessing_threads 4 \
    --num_readers 8 \
    --moving_average_decay=0.9999 \
    --weight_decay=0.00005 \
    --learning_rate_decay_type=exponential \
    --learning_rate_decay_factor=0.97 \
    --label_smoothing=0.1
```

**Other info / logs**
I thought this issue #20843 is related to the problem but I have done a git bisect between the `v1.8.0` and the `v1.9.0` tag and running the script above for every commit and found the one that introduce the largest drop in performance: d4976f7. This commit enabled the layout optimizer by default for all gpu cluster. As a workaround and a proof I disabled the layout optimizer with the following config:
```python
rewrite_options = rewriter_config_pb2.RewriterConfig(layout_optimizer=rewriter_config_pb2.RewriterConfig.OFF)
graph_options = tf.GraphOptions(rewrite_options=rewrite_options)
config = tf.ConfigProto(graph_options=graph_options) 
```  
And most of the performance were back again in `tensorflow 1.9.0-devel-gpu-py3`:
`INFO:tensorflow:global_step/sec: 1.18333`
And in the latest release `tensorflow 1.12.0-devel-gpu-py3`:
`INFO:tensorflow:global_step/sec: 1.23337`
[Here](https://gist.github.com/DavidWiesner/c6af3238a90b3a8b1209543a9f8d7127#file-git-bisect-log) is a full log of my bisect script. [One another log](https://gist.github.com/DavidWiesner/c6af3238a90b3a8b1209543a9f8d7127#file-git-bisect-layout_optimizer-on-off-log) shows the influence of disabling the layout-optimizer in commit d4976f7 (that introducing the drop in performance) and some release versions of tensorflow. 
Maybe @yaozhang can answer why the layout optimizer was disabled for some architectures until 1.8.0 in f0d1abbf2 and why @zhangyaobit enabled for all architectures since 1.9.0 in d4976f7

**Comment by ezhulenev** [comment]
I was able to measure 10-15% performance regression with a batch size 4 on a single V100 GPU (larger batch OOMs). V100 has faster (30%-50%) convolutions (and conv gradients) in NCHW data format, and layout optimizer correctly swaps all convs from default NHWC to NCHW in this model, however it seems that all the extra transpose nodes are killing all the conv gains.

Looking at the executed graph and profiles, maybe layout optimizer missed an opportunity to remove redundant transposes.

**Comment by ezhulenev** [comment]
The problem is that NASNet model has mostly 1x1 convolutions, and in NHWC format they are computed as a matrix multiplication, in NCHW they go though cuDNN convolution, and it's ~30-50% slower. I'll try to come up with a better strategy for layout swapping,  though it might be problematic, because currently we can only have a single "target" data format for the graph.

**Comment by byronyi** [comment]
@ezhulenev Any updates on this?

**Comment by ezhulenev** [comment]
I'm working on a new layout optimizer in MLIR framework, and this problem is solved there, however I'm not sure when it will be on by default, should have some certainty in a few weeks.

**Comment by rmothukuru** [comment]
@DavidWiesner,
 Sorry for the delayed response. As per [this comment](https://github.com/tensorflow/tensorflow/issues/23847#issuecomment-604629618), your issue should be resolved. Can you please check and let us know if it is so? Thanks!
